File: SparkQueries/WordCount.scala
import org.apache.hadoop.fs.{FileSystem, Path}
import java.lang.management.ManagementFactory
import java.lang.management.MemoryMXBean

// Load the data from the CSV file
val linesRDD = sc.textFile("hdfs:///user/cloudera/data/final_dataset.csv")

// Start timer for execution time measurement
val startTime = System.nanoTime()

// Skip the header row and split words
val header = linesRDD.first()
val dataRDD = linesRDD.filter(row => row != header)

// Perform word count
val wordsRDD = dataRDD.flatMap(line => line.split("[,\\s]+").map(_.trim)) // Split on commas and whitespace
val wordCountRDD = wordsRDD
  .filter(_.nonEmpty) // Exclude empty strings
  .map(word => (word, 1)) // Map each word to 1
  .reduceByKey(_ + _) // Aggregate counts
  .sortBy(_._2, ascending = false) // Sort by count in descending order

// Format the output as "word, count"
val formattedOutputRDD = wordCountRDD.map { case (word, count) => s"$word, $count" }

// Define the output path
val outputPath = "hdfs:///user/cloudera/data/output_wordcount"
val fs = FileSystem.get(sc.hadoopConfiguration)
val path = new Path(outputPath)

// Check if output path exists and delete it if necessary
if (fs.exists(path)) {
  fs.delete(path, true)
}

// Save the results to HDFS
formattedOutputRDD.saveAsTextFile(outputPath)

// End timer and calculate duration
val endTime = System.nanoTime()
val durationInSeconds = (endTime - startTime) / 1e9d

// Count the total records processed
val totalRecords = dataRDD.count()

// Calculate throughput
val throughput = totalRecords / durationInSeconds

// Measure memory usage
val runtime = Runtime.getRuntime
val usedMemory = (runtime.totalMemory - runtime.freeMemory) / (1024.0 * 1024.0)
val memoryMXBean = ManagementFactory.getMemoryMXBean
val heapMemoryUsed = memoryMXBean.getHeapMemoryUsage.getUsed / (1024.0 * 1024.0)
val nonHeapMemoryUsed = memoryMXBean.getNonHeapMemoryUsage.getUsed / (1024.0 * 1024.0)

// Print metrics
println(s"Execution Time: $durationInSeconds seconds")
println(s"Throughput: $throughput records per second")
println(s"Used Memory: $usedMemory MB")
println(s"Heap Memory Used: $heapMemoryUsed MB")
println(s"Non-Heap Memory Used: $nonHeapMemoryUsed MB")

println(s"Output saved to: $outputPath")

File: HadoopQueries/WordCount.java
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

public class WordCount {

    public static enum Counter {
        RECORD_COUNT
    }

    public static class TokenizerMapper extends Mapper<Object, Text, Text, IntWritable>{
         private final static IntWritable one = new IntWritable(1);
         private Text word = new Text();
         private int lineNumber = 0;

         public void map(Object key, Text value, Context context) throws IOException, InterruptedException {
            // Skip the header row
            if (lineNumber == 0) {
                lineNumber++;
                return;
            }
           String[] fields = value.toString().split(",");

            // Process each field as a word
            for (String field : fields) {
                word.set(field.trim());
                context.write(word, one);
            }
            lineNumber++;
            context.getCounter(Counter.RECORD_COUNT).increment(1);
         }
    }

    public static class IntSumReducer extends Reducer<Text,IntWritable,Text,IntWritable> {
         private IntWritable result = new IntWritable();

         public void reduce(Text key, Iterable<IntWritable> values, Context context) throws IOException, InterruptedException {
             int sum = 0;
             for (IntWritable val : values) {
                  sum += val.get();
             }
             result.set(sum);
             context.write(key, result);
         }
    }



    public static void main(String[] args) throws Exception {
        long startTime = System.nanoTime();
        Configuration conf = new Configuration();
        Job job = Job.getInstance(conf, "word count");
        job.setJarByClass(WordCount.class);
        job.setMapperClass(TokenizerMapper.class);
        job.setCombinerClass(IntSumReducer.class);
        job.setReducerClass(IntSumReducer.class);
        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(IntWritable.class);
        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileOutputFormat.setOutputPath(job, new Path(args[1]));

        boolean jobStatus = job.waitForCompletion(true);
        long endTime = System.nanoTime();
        double duration = (endTime - startTime) / 1e9d;
        System.out.println("Execution Time: " + duration + " seconds");
        long recordCount = job.getCounters().findCounter(Counter.RECORD_COUNT).getValue();
        double throughput = recordCount / duration;
        System.out.println("Throughput: " + throughput + " records per second");

        Runtime runtime = Runtime.getRuntime();
        double usedMemory = (runtime.totalMemory() - runtime.freeMemory()) / (1024.0 * 1024.0);
        System.out.println("Used Memory: " + usedMemory + " MB");

        MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
        double usedHeapMemory = memoryMXBean.getHeapMemoryUsage().getUsed() / (1024.0 * 1024.0);
        double usedNonHeapMemory = memoryMXBean.getNonHeapMemoryUsage().getUsed() / (1024.0 * 1024.0);
        System.out.println("Used Heap Memory: " + usedHeapMemory + " MB");
        System.out.println("Used Non-Heap Memory: " + usedNonHeapMemory + " MB");

        System.exit(jobStatus ? 0 : 1);
     }
}
